Backports release 1.10

KristofferC opened the pull request:
Backported PRs:
- [x] #51234 <!-- Fix getfield codegen for tuple inputs and unknown symbol fields. -->
- [x] #52170 <!-- fix invalidations related to `ismutable` -->
- [x] #52342 <!-- Add single-term multiplication for `AbstractQ` on v1.10 and above -->
- [x] #52333 <!-- bugfix for dot of Hermitian{noncommutative} -->
- [x] #52407 <!-- channels: fix memory ordering violation in iterate -->
- [x] #52405 <!-- Bump LLVM to 15.0.7+10 to fix GC issue -->
- [x] #52441 <!-- Remove `Pkg` dependency from `SuiteSparse_jll` -->
- [x] #52367 <!-- docs: add notes about scratchspaces in depot -->
- [x] #52456 <!-- Make `jl_write_coverage_data` dllexported again -->
- [x] #52294 <!-- GC scheduler refinements -->
- [x] #52359 <!-- make custom log macros work -->
- [x] #52548 

Need manual backport:

Contains multiple commits, manual intervention needed:
- [ ] #51092 <!-- inference: fix bad effects for recursion -->

Non-merged PRs with backport label:
- [ ] #52447 <!-- fix isequal_normalized for combining-char reordering -->
- [ ] #52362 <!-- Document environment variable JULIA_PKG_PRESERVE_TIERED_INSTALLED -->
- [ ] #52354 <!-- Fix `show` for `MethodList` when methods are from another module -->
- [ ] #51479 <!-- prevent code loading from lookin in the versioned environment when building Julia -->

ranocha committed:
fix invalidations related to `ismutable` (#52170)

Related to #52134. It would be nice if the underlying inference issue
was fixed but this seems to be a hotfix for now.

I have seen this inference problem occurring in Julia v1.9, v1.10, and
current `master`. For example, on Julia v1.9.3, I get

```julia

julia> code_warntype(ismutable, (Function,))
MethodInstance for ismutable(::Function)
  from ismutable(x) @ Base reflection.jl:521
Arguments
  #self#::Core.Const(ismutable)
  x::Function
Body::Any
1 ─      nothing
│        nothing
│   %3 = Base.typeof(x)::Type{<:Function}
│   %4 = Base.getproperty(%3, :name)::Any
│   %5 = Base.getproperty(%4, :flags)::Any
│   %6 = (%5 & 0x02)::Any
│   %7 = (%6 == 0x02)::Any
└──      return %7
```

This causes some invalidations when `using OrdinaryDiffEq`.

(cherry picked from commit df40bab2dadf6ad46b822e5ea3e71ac3ee238628)

IanButterworth committed:
make custom log macros work (#52359)

(cherry picked from commit 641f717a15cf89964db446c1bb32d0d672763786)

jishnub committed:
Add single-term multiplication for `AbstractQ` on v1.10 and above (#52342)

(cherry picked from commit 150c1adfd6f1a54e6a74393c7bb73af8f772fa30)

stevengj committed:
bugfix for dot of Hermitian{noncommutative} (#52333)

Co-authored-by: Daniel Karrasch <daniel.karrasch@posteo.de>
(cherry picked from commit 53f1eb82b8a4265def61ab677746b63a583ef865)

d-netto committed:
GC scheduler refinements (#52294)

Supersedes https://github.com/JuliaLang/julia/pull/51061 and
https://github.com/JuliaLang/julia/pull/51414.

Still needs more perf analysis.

(cherry picked from commit e26c257b28ec6285c8687aef466b64f9ce23c3c7)

vtjnash committed:
channels: fix memory ordering violation in iterate (#52407)

Channel `iterate` calls might miss trailing items without this patch. I
have not seen proof of this reaching a failure, but we do appear to be
missing this ordering specification in visual review.

Should not make a difference to generated code on x86, which already has
TSO guaranteed, but may alter optimizations and other CPUs with weaker
memory orderings.

(cherry picked from commit 856e1120a8f255e4907a05f0146a2026a6665dd4)

gbaraldi committed:
Bump LLVM to 15.0.7+10 to fix GC issue (#52405)

(cherry picked from commit e5046b4579cf571931714abbe14a3a049ca6383b)

hycakir committed:
Remove `Pkg` dependency from `SuiteSparse_jll` (#52441)

This dependency seems to be superfluous, and `Pkg` brings lots of
dependencies into the project (Downloads, LibGit2, etc.) which makes it
hard to create leaner binary packages via `PackageCompiler` for projects
that depend on `SuiteSparse_jll`.

(cherry picked from commit 3b9c39ac1cbb46688d7b3597e70b9ae591e3b0ec)

IanButterworth committed:
docs: add notes about scratchspaces in depot (#52367)

Co-authored-by: Fredrik Ekre <ekrefredrik@gmail.com>
(cherry picked from commit bf6c31ceb5c4da5839026463bc29d00458d80a75)

Drvi committed:
Make `jl_write_coverage_data` dllexported again (#52456)

Closes #52452

(cherry picked from commit 8cf7598ebd0c120f50db2f5d6bbd091855a4836c)

KristofferC commented:
@nanosoldier `runtests()`

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/52503#issuecomment-1855358408) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/c61b27f_vs_0ba6ec2/report.html) is available.

aviatesk committed:
fix #52531, fix the effects modeling of `QuoteNode` (#52548)

What observed in #52531 is that `QuoteNode` can embed global variables
that users can modify. Therefore, when dealing with `QuoteNode`, it's
necessary to taint its `:inaccessiblememonly` just like we do for
`GlobalRef`.

- fixes #52531
- replaces #52536
